src/wallet/wallet.cpp

```diff
@@ -2956,11 +2952,21 @@ bool CWallet::CreateTransactionInternal(
                     txNew.vout.push_back(txout);
                 }
 
+                // Include the fees for things that aren't inputs, excluding the change output
+                const CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
+                CAmount nValueToSelect = nValue + not_input_fees;
+
+                // For KnapsackSolver, when we are not subtracting the fee from the recipients, we also want to include the fees for the
+                // inputs that we found in the previous iteration.
+                if (!coin_selection_params.use_bnb && nSubtractFeeFromAmount == 0) {
+                    nValueToSelect += std::max(CAmount(0), nFeeRet - not_input_fees);
+                }
+
```

Comment:
In commit "Roll static tx fees into nValueToSelect instead of having it be separate" (67fc8b7be1d5fbdb199d6e995434431a7282d446)

This logic is not making sense to me. It wasn't this complicated before, and even though commits are reordered and nTargetValue is passed differently now I would expect these lines 2957-2964 to look like:

```c++
CAmount nValueToSelect = nValue;
if (coin_selection_params.use_bnb) nValueToSelect += not_input_fees;
if (nSubtractFeeFromAmount != 0) nValueToSelect += nFeeRet;
```

This would:
- Avoid adding not_input_fees and then subtracting it
- Avoid introducing unexplained std::max not_input_fees > nFeeRet corner case
- Avoid changing behavior in the bnb && nSubtractFeeFromAmount case.
  - If it is actually desirable to change bnb behavior I think it should happen in a different commit. This commit is already doing multiple things and I think would be good to split, see https://github.com/bitcoin/bitcoin/pull/17331#discussion_r632069270
- Let you drop or shorten the "Note that setting nValueToSelect is changed..." paragraph in commit description since it should be obvious this only changes bnb

Reply:
I had done it this way so that a future commit which has to remove this special case would not need to modify any lines, just drop the code. This code is required at this commit because nFeeRet itself includes not_input_fees, so KnapsackSolver would always be trying to choose more coins than it actually needs. However in the next commit, this is just dropped directly because it is no longer needed once KnapsackSolver uses effective value.